### PR TITLE
Open a new file on double-clicking the tab bar

### LINKED ISSUE #6818: Double-Click to Open New Files
### Check for existing issues

- [X] Completed

### Describe the feature

A nice feature to have would be to allow double clicking on the home screen, or the tabs bar, to create new files:

https://github.com/zed-industries/zed/assets/5413308/a863e5a1-42cb-444d-905e-cfc62272b4f4

### If applicable, add mockups / screenshots to help present your vision of the feature

_No response_

### PR DESCRIPTION
**Pull request by hmemcpy**
Fixes #6818

This is a convenience feature that exists in other editors, allowing quick creation a new tab by double-clicking the empty space on the tab bar:

Release Notes:

- Added the ability to open a new buffer when double-clicking on the tab bar ([#6818](https://github.com/zed-industries/zed/issues/6818)).

### CODE REVIEW

File: crates/ui/src/components/tab_bar.rs [hunk 1]
```diff
@@ -1,11 +1,10 @@
-use gpui::{AnyElement, ScrollHandle};
-use smallvec::SmallVec;
-
 use crate::prelude::*;
+use gpui::{AnyElement, ScrollHandle, Stateful};
```
Comment (maxdeviant) [inline crates/ui/src/components/tab_bar.rs:2]: Let's keep these imports where they were.

We want to separate third-party imports from local `crate::` import blocks.

File: crates/ui/src/components/tab_bar.rs [hunk 2]
```diff
@@ -88,12 +88,29 @@ impl ParentElement for TabBar {
...
+        handler: impl Fn(&gpui::ClickEvent, &mut WindowContext) + 'static,
+    ) -> Self {
+        self.interactivity().on_click(handler);
+        self
+    }
+}
+
 impl RenderOnce for TabBar {
-    fn render(self, cx: &mut WindowContext) -> impl IntoElement {
+    #[allow(refining_impl_trait)]
```
Comment (maxdeviant) [inline crates/ui/src/components/tab_bar.rs:109]: Why do we need to refine the trait here?

File: crates/ui/src/components/tab_bar.rs (hunk 1 above)
Comment (hmemcpy) [reply crates/ui/src/components/tab_bar.rs:2]: I ran rustfmt and it did that, but you're right - I'll revert this to the previous order.

File: crates/ui/src/components/tab_bar.rs (hunk 2 above)
Comment (hmemcpy) [reply crates/ui/src/components/tab_bar.rs:109]: Without it, I got a compiler warning. I saw that some other types that expose `RenderOnce` added this refinement. I'm very new to Rust, so I don't have a better reason 😅

**Comment by maxdeviant** [comment]
I think this probably belongs on the same element we use for the tab drop target:

https://github.com/zed-industries/zed/blob/a44fc24445809c21cf56150b1031e624ea99f7ec/crates/workspace/src/pane.rs#L1507-L1531

This element takes up the remaining space not occupied by tabs in the tab bar, so we won't have to worry about the click handlers interfering:

Screenshot 2024-02-26 at 11 42 21 AM

It looks like there's currently some sort of flickering issue with the element (I suspect it's related to the other ones we're looking into), so just be mindful of that when implementing.

**Review by maxdeviant — CHANGES_REQUESTED** [review]
See my comment above.

I don't think this functionality belongs on the `TabBar` itself.

**Comment by hmemcpy** [comment]
Ah, very cool. I'll take another look. Thanks for the comments!

**Comment by hmemcpy** [comment]
@maxdeviant hmm, for some reason I can't get the mouse events to respond when adding it to the drop target element. I tried bringing over the definition of `on_click` from the `StatefulInteractiveElement` to `InteractiveElement`, as well as using the raw `on_mouse_down` already available, but they both don't seem to work for me - I never get into the listener callback.

I might be holding it wrong :) Any ideas?

**Comment by maxdeviant** [comment]
> @maxdeviant hmm, for some reason I can't get the mouse events to respond when adding it to the drop target element. I tried bringing over the definition of `on_click` from the `StatefulInteractiveElement` to `InteractiveElement`, as well as using the raw `on_mouse_down` already available, but they both don't seem to work for me - I never get into the listener callback.
> 
> I might be holding it wrong :) Any ideas?

Try slapping a background color (e.g., `debug_bg_cyan()`) on it to make sure the element is where you expect it to be.

The aforementioned flickering issue causes the width of it to move back and forth, so that could be interfering with it's location or its events.

**Comment by hmemcpy** [comment]
Thanks for the hint, it does indeed flicker. Before changing the background, I was able to observe this by adding a println inside the `on_mouse_down` function, and it prints all the time - something must be continuously redrawing all components causing it to re-register those events, maybe?

Finally, `on_mouse_down` works, but it doesn't capture the `click_count`, which I suppose is why the `on_click` is only implemented on the `StatefulInteractiveElement`. Is it possible to make `on_click` work with the `InteractiveElement` instead?

**Review by maxdeviant — APPROVED** [review]
Just tested it out and works great!

**Pull request merged**